deeplearning4j-zoo/src/main/java/org/deeplearning4j/zoo/model/FaceNetNN4Small2.java

```diff
@@ -26,27 +29,18 @@
  *
  * Revised and consolidated version by @crockpotveggies
  */
-@NoArgsConstructor
+@AllArgsConstructor
+@Builder
 public class FaceNetNN4Small2 extends ZooModel {
 
-    private int[] inputShape = new int[] {3, 96, 96};
+    @Builder.Default private long seed = 1234;
+    @Builder.Default private int[] inputShape = new int[] {3, 96, 96};
     private int numLabels;
-    private long seed;
-    private Activation transferFunction = Activation.RELU;
-    private WorkspaceMode workspaceMode;
-    private ConvolutionLayer.AlgoMode cudnnAlgoMode;
-
-    public FaceNetNN4Small2(int numLabels, long seed) {
-        this(numLabels, seed, WorkspaceMode.ENABLED);
-    }
-
-    public FaceNetNN4Small2(int numLabels, long seed, WorkspaceMode workspaceMode) {
-        this.numLabels = numLabels;
-        this.seed = seed;
-        this.workspaceMode = workspaceMode;
-        this.cudnnAlgoMode = workspaceMode == WorkspaceMode.ENABLED ? ConvolutionLayer.AlgoMode.PREFER_FASTEST
-                        : ConvolutionLayer.AlgoMode.NO_WORKSPACE;
-    }
+    @Builder.Default private IUpdater updater = new Adam(0.1, 0.9, 0.999, 0.01);
+    @Builder.Default private Activation transferFunction = Activation.RELU;
+    @Builder.Default CacheMode cacheMode = CacheMode.DEVICE;
```

Comment:
This applies to all nets: I'm not comfortable with this being the default, when it's *not* the default for DL4J in general. I haven't run the benchmarks here, but I do know cache mode can use a *lot* of memory... for each conv layer, the cached im2col array is ```activationsNumElements * kH * kW / (sH * sW)```. For something like the early layers of AlexNet (11x11 / 4x4 =7.56 if memory serves me) that's a lot of memory.

Reply:
@raver119 can you comment here? personally, I've seen negligible memory usage when cache mode is enabled on VGG16 (with the exception of that strange OOM on batch size > 24).

Reply:
Because cache mode in my own experience has consistently led to good performance, I'm not in favour of changing the default. Taking a quick look at nvidia-smi on VGG-16, here's the memory usage of cache mode enabled vs disabled. That's not enough for my concern, but please tell me otherwise:

| Model | Cache Mode | Batch | Memory Usage |
|---|---|---|---|
| VGG16 | Enabled | 16 | 3733MiB / 12194MiB |
| VGG16 | Disabled | 16 | 3065MiB / 12194MiB |
| AlexNet | Enabled | 48 | 2317MiB / 12194MiB |
| AlexNet | Disabled | 48 | 2193MiB / 12194MiB |

Reply:
This doesn't apply just to the cache setting though. There's a lot of configuration that we cannot change with this interface, so I've always considered what actually gets set as default to be a moot point. We often end up just calling conf() or something and taking it up from there.

Reply:
@crockpotveggies Was CuDNN used for these tests?

Reply:
Yes cuDNN was used in these tests.

Reply:
CacheMode doesn't apply when using CuDNN - it can't, due to how cudnn works

Note the CuDNN helper call (and return) happens before the caching:
https://github.com/deeplearning4j/deeplearning4j/blob/491507004a571ba0793fac90548a1d8515fd8e63/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/convolution/ConvolutionLayer.java#L306-L336

https://github.com/deeplearning4j/deeplearning4j/blob/491507004a571ba0793fac90548a1d8515fd8e63/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/convolution/ConvolutionLayer.java#L372-L376

Reply:
Doesn't make sense here. All of those values I showed you used cuDNN. So what is making the memory difference?

Reply:
Not sure, there shouldn't be any difference. Open an issue if you're confident there is a difference here?
(Though I'd recommend checking/confirming it with the memory benchmarks first)